File: backend/app/schemas/storyboard.py
"""Storyboard API schemas (Pydantic v2)."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.storyboard import ImagesStatus, StoryboardStatus


# ── Request ───────────────────────────────────────────────────

class GenerateRequest(BaseModel):
    scene_description: str = Field(
        ...,
        min_length=20,
        max_length=2000,
        description="Plain-English description of the scene (20–2000 chars).",
    )
    scene_style: str = Field(
        default="Cinematic",
        max_length=50,
        description="Visual mood: Cinematic, Moody, Warm, Dramatic, Vintage, Cold …",
    )
    location_desc: str | None = Field(
        default=None,
        max_length=500,
        description="Optional location description for richer shot suggestions.",
    )
    character_desc: str | None = Field(
        default=None,
        max_length=500,
        description="Optional character descriptions.",
    )


# ── Inner response ────────────────────────────────────────────

class ShotOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    shot_number: int
    shot_name: str | None
    shot_type: str | None
    camera_angle: str | None
    camera_movement: str | None
    lens: str | None
    lighting: str | None
    mood: str | None
    duration: str | None
    prompt: str | None
    explanation: str | None
    image_url: str | None


class StoryboardOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str | None
    scene_description: str
    scene_style: str | None
    director_note: str | None
    status: StoryboardStatus
    images_status: ImagesStatus
    shots: list[ShotOut]
    created_at: datetime


class StoryboardSummary(BaseModel):
    """Lightweight item for the history list."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str | None
    scene_style: str | None
    status: StoryboardStatus
    shot_count: int = 0
    created_at: datetime

File: mobile/src/services/storyboard.ts
/**
 * Storyboard API calls — thin wrappers over the generic `api` client.
 */

import { api } from "./api";

export interface ShotOut {
  id: string;
  shot_number: number;
  shot_name: string | null;
  shot_type: string | null;
  camera_angle: string | null;
  camera_movement: string | null;
  lens: string | null;
  lighting: string | null;
  mood: string | null;
  duration: string | null;
  prompt: string | null;
  explanation: string | null;
  image_url: string | null;
}

export interface StoryboardOut {
  id: string;
  title: string | null;
  scene_description: string;
  scene_style: string | null;
  director_note: string | null;
  status: "pending" | "processing" | "completed" | "failed";
  shots: ShotOut[];
  created_at: string;
}

export interface StoryboardSummary {
  id: string;
  title: string | null;
  scene_style: string | null;
  status: "pending" | "processing" | "completed" | "failed";
  shot_count: number;
  created_at: string;
}

export interface GenerateRequest {
  scene_description: string;
  scene_style?: string;
  location_desc?: string | null;
  character_desc?: string | null;
}

export const storyboardApi = {
  generate: (body: GenerateRequest) =>
    api.post<StoryboardOut>("/storyboard/generate", body),

  history: (limit = 20, offset = 0) =>
    api.get<StoryboardSummary[]>(`/storyboard/history?limit=${limit}&offset=${offset}`),

  getOne: (id: string) =>
    api.get<StoryboardOut>(`/storyboard/${id}`),

  delete: (id: string) =>
    api.del<{ deleted: boolean }>(`/storyboard/${id}`),
};
